### ISSUE TITLE
Compatibility with Linux/amd64

### ISSUE THREAD

**Issue by Stevenlimo** [anvil-works/anvil-runtime#14]
Hi administrator, 
I tried to test the default anvil-app with a super user name 'tzuser', but it gave me the following error:

-------------------------------------------------------------------------------------
(venv) tzuser@mysite:~/venv$ sudo anvil-app-server --app MyTodoList
Found Anvil App Server JAR in package directory
Failed to start built-in Postgres database: java.lang.IllegalStateException: Process [.anvil-data/db-bin/PG-a8263178df5b3591feca6b619939c2bb/bin/initdb, -A, trust, -U, postgres, -D, .anvil-data/db, -E, UTF-8] failed
More logs are available in .anvil-data/postgres.log.
Some common causes of this problem:
 - Are you launching this server as 'root' on a UNIX system?
   Postgres will not run as root; try launching the server as an ordinary user.
 - Are you running this server on an unusual architecture or OS? (Linux/amd64)
-------------------------------------------------------------------------------------

 Is it something to do with my VPS architecture, which is follow:
-------------------------------------------------------------------------------------
(venv) tzuser@mysite:~/venv$ uname -a
Linux mysite 4.15.0-20-generic #21-Ubuntu SMP Tue Apr 24 06:16:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
-------------------------------------------------------------------------------------

Please advice, thanks.

**Comment by andersgs** [comment]
Hi @Stevenlimo.

The answer is right there in the output:

```
Are you launching this server as 'root' on a UNIX system?
Postgres will not run as root; try launching the server as an ordinary user.
```

Remove the `sudo` from your command:

```
anvil-app-server --app MyTodoList
```

Cheers. A.

**Comment by Stevenlimo** [comment]
Thank you I did, but it came with the same error and more as follow:
---------------------------------------------------------------------------
tzuser@mysite:~$ cd venv
tzuser@mysite:~/venv$ source bin/activate
(venv) tzuser@mysite:~/venv$ clear
(venv) tzuser@mysite:~/venv$ anvil-app-server --app MyTodoList
Found Anvil App Server JAR in package directory
Failed to start built-in Postgres database: java.io.FileNotFoundException: /home/tzuser/venv/.anvil-data/postgres-embedding.log (Permission denied)
More logs are available in .anvil-data/postgres.log.
Some common causes of this problem:
 - Are you launching this server as 'root' on a UNIX system?
   Postgres will not run as root; try launching the server as an ordinary user.
 - Are you running this server on an unusual architecture or OS? (Linux/amd64)
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Parser/parser.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/structmember.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/mystrtoul.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/getcopyright.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/getcompiler.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/ast_unparse.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/ast_opt.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
------------------------------------------------------------------------------------------

Any ideas?

**Comment by andersgs** [comment]
Is `tzuser` a super user, perhaps?

My suggestion would be to create an ordinary user (e.g., called `anvil`), change to that user (`sudo -i -u anvil`), and then try again.

I would not give the `anvil` user the ability to login via ssh: https://www.cyberciti.biz/tips/howto-linux-shell-restricting-access.html

Good luck.

Also of possible interest: https://www.cyberciti.biz/faq/create-a-user-account-on-ubuntu-linux/

**Comment by Stevenlimo** [comment]
I removed the 'tzuser' from the Sudo group, so now it is an ordinary user:
----------------------------------------------------------------------------------
(venv) tzuser@mysite:~/venv$ sudo anvil-app-server --app MyTodoList
tzuser is not in the sudoers file.  This incident will be reported.
----------------------------------------------------------------------------------

Then I tried to run again, the same errors repeat
----------------------------------------------------------------------------------
root@mysite:~# sudo -i -u tzuser
(venv) tzuser@mysite:~/venv$ anvil-app-server --app MyTodoList
Found Anvil App Server JAR in package directory
Failed to start built-in Postgres database: java.io.FileNotFoundException: /home/tzuser/venv/.anvil-data/postgres-embedding.log (Permission denied)
More logs are available in .anvil-data/postgres.log.
Some common causes of this problem:
 - Are you launching this server as 'root' on a UNIX system?
   Postgres will not run as root; try launching the server as an ordinary user.
 - Are you running this server on an unusual architecture or OS? (Linux/amd64)
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Parser/parser.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/structmember.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/mystrtoul.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/getcopyright.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/getcompiler.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/ast_unparse.gcda:Skip
profiling:/root/Python-3.7.5:Cannot create directory
profiling:/root/Python-3.7.5/Python/ast_opt.gcda:Skip
----------------------------------------------------------------------------------

I tried with different user names but it all came with the same error. I am not sure it is the Linux architecture problem or not.

**Comment by andersgs** [comment]
It seems it is trying to install stuff in this folder `/root/Python-3.7.5/`. What are the permissions on that folder? Did you install `app-server` in the virtual environment?

What is the output of `which anvil-app-server`?

**Comment by Stevenlimo** [comment]
Yes,  the app is in the virtualenv folder 'venv' in the tzuser account. The main Python3.7 program is installed in the root account. I can only use 'Sudo' in my tzuser account to avoid this permission problem as shown in my first post. I don't really know what went wrong from there.

**Comment by andersgs** [comment]
You may want to bring this up with your VPS provider.

Two other options to consider:
* install miniconda in your home directory: https://docs.conda.io/en/latest/miniconda.html. This will provide you with a local to your user Python interpreter, and you can then install the app-server from there.

* run the app server from a Docker container as a non-privileged user within the container.

Maybe the Anvil devs have a better idea. Sorry I was not more help.

**Comment by Stevenlimo** [comment]
Thank you for your great help. I will keep trying.

**Comment by andersgs** [comment]
I assume you followed the instructions here: https://github.com/anvil-works/anvil-runtime/blob/master/doc/getting-started.md#cookbook-for-serving-an-anvil-app-from-a-public-linux-server

**Comment by JackN-code** [comment]
I seems that you VPS has low RAM 512, please try 1G. I had the same issue before

**Comment by Stevenlimo** [comment]
Thank you gents. I managed to get the app running,

**Comment by JackN-code** [comment]
Would you mind tell what is the issue and how do you fix it?

**Comment by Stevenlimo** [comment]
Yes, the issue is: after I terminated the anvil-app-server, and run the server again, the port 3030 was still running in the background, so I had to kill the port 3030 process in order for the next run to be successful. Sometimes, I also need to kill the postgres process and to restart it again from the root. That was what happened to my problem. Everytime I run the anvil-app-server, I have to do the above two process killing before running the server again. Hope this help.
